pkg/capabilities/cli/cmd/generated_info.go

```diff
@@ -0,0 +1,173 @@
+	Input          *Struct
+	Types          map[string]Struct
+	CapabilityType capabilities.CapabilityType
+	BaseName       string
+	RootOutput     string
+	RootNumSlice   int
+	ExtraImports   []string
+	ID             *string
+}
+
+func (g GeneratedInfo) RootType() Struct {
+	if r, ok := g.Types[g.RootOutput]; ok {
+		return r
+	}
+	return Struct{
+		Name: g.RootOutput,
+		Ref:  &g.RootOutput,
+	}
+}
+
+func generatedInfoFromSrc(src string, capID *string, typeInfo TypeInfo) (GeneratedInfo, error) {
+	fset := token.NewFileSet()
+
+	// Parse the source code string
+	node, err := parser.ParseFile(fset, "", src, parser.AllErrors)
+	if err != nil {
+		return GeneratedInfo{}, err
+	}
+	pkg := node.Name.Name
+
+	generatedStructs := map[string]Struct{}
+	var extraImports []string
+	ast.Inspect(node, func(n ast.Node) bool {
+		return inspectNode(n, fset, src, generatedStructs, &extraImports)
+	})
+
+	root := generatedStructs[typeInfo.RootType]
+	input, config := extractInputAndConfig(generatedStructs, typeInfo, root)
+
+	output := root.Outputs["Outputs"]
+
+	return GeneratedInfo{
+		Package:        pkg,
+		Config:         config,
+		Types:          generatedStructs,
+		RootOutput:     output.Type,
+		RootNumSlice:   output.NumSlice,
+		BaseName:       typeInfo.RootType,
+		CapabilityType: typeInfo.CapabilityType,
+		Input:          input,
+		ExtraImports:   extraImports,
+		ID:             capID,
+	}, nil
+}
+
+func extractInputAndConfig(generatedStructs map[string]Struct, typeInfo TypeInfo, root Struct) (*Struct, Struct) {
+	delete(generatedStructs, typeInfo.RootType)
+	inputField, ok := root.Outputs["Inputs"]
+	var input *Struct
+	if ok {
+		inputType := inputField.Type
+		inputS, ok2 := generatedStructs[inputType]
+		if ok2 {
+			input = &inputS
+			delete(generatedStructs, inputType)
+		} else {
+			input = &Struct{
+				Name: lastAfterDot(inputType),
+				Ref:  &inputType,
+			}
+		}
+	}
+
+	configType := root.Outputs["Config"].Type
+	config, ok := generatedStructs[configType]
+	if !ok {
+		config = Struct{
+			Name: lastAfterDot(configType),
+			Ref:  &configType,
+		}
+	}
+	for k := range generatedStructs {
+		if strings.HasPrefix(k, configType) || (input != nil && strings.HasPrefix(k, input.Name)) {
+			delete(generatedStructs, k)
+		}
+	}
+	return input, config
+}
+
+func inspectNode(n ast.Node, fset *token.FileSet, src string, rawInfo map[string]Struct, extraImports *[]string) bool {
+	if ts, ok := n.(*ast.TypeSpec); ok {
+		s := Struct{
+			Name:    strings.TrimSpace(ts.Name.Name),
+			Outputs: map[string]Field{},
+		}
+
+		if structType, ok := ts.Type.(*ast.StructType); ok {
+			for _, field := range structType.Fields.List {
+				start := fset.Position(field.Type.Pos()).Offset
+				end := fset.Position(field.Type.End()).Offset
+				typeStr := src[start:end]
+				if typeStr == "interface{}" {
+					typeStr = "any"
+				}
+				f := Field{}
+
+				if field.Tag != nil {
+					tag := reflect.StructTag(field.Tag.Value[1 : len(field.Tag.Value)-1])
```

Comment:
this looks like it can panic. why is structTag guaranteed to have at least 1 elem?

Reply:
The way we called the generator for the other code will guarantee it.  I can add a comment to that effect.